core/node/node_framework/src/service/tests.rs

```diff
@@ -0,0 +1,235 @@
+}
+
+// ZkStack Service's `run()` method has to take into account errors on wiring step.
+#[derive(Debug)]
+struct TaskErrorLayer;
+
+#[async_trait::async_trait]
+impl WiringLayer for TaskErrorLayer {
+    fn layer_name(&self) -> &'static str {
+        "task_error_layer"
+    }
+
+    async fn wire(self: Box<Self>, mut node: ServiceContext<'_>) -> Result<(), WiringError> {
+        node.add_task(Box::new(ErrorTask));
+        Ok(())
+    }
+}
+
+#[derive(Debug)]
+struct ErrorTask;
+
+#[async_trait::async_trait]
+impl Task for ErrorTask {
+    fn name(&self) -> &'static str {
+        "error_task"
+    }
+    async fn run(self: Box<Self>, _stop_receiver: StopReceiver) -> anyhow::Result<()> {
+        anyhow::bail!("error task")
+    }
+}
+
+// ZkStack Service's `run()` method has to take into account errors inside task execution.
+#[test]
+fn test_run_with_failed_tasks() {
+    let mut zk_stack_service: ZkStackService = ZkStackService::new().unwrap();
+    zk_stack_service.add_layer(TaskErrorLayer);
+    let res = zk_stack_service.run();
+    assert_eq!(
+        res.unwrap_err().to_string(),
+        "Task error_task failed".to_string()
+    );
+}
+
+// Layer that adds a task that updates resource and returns
+#[derive(Debug)]
+struct SuccessfulTaskLayer(Arc<Mutex<bool>>);
+
+#[async_trait::async_trait]
+impl WiringLayer for SuccessfulTaskLayer {
+    fn layer_name(&self) -> &'static str {
+        "successful_task_layer"
+    }
+
+    async fn wire(self: Box<Self>, mut node: ServiceContext<'_>) -> Result<(), WiringError> {
+        node.add_task(Box::new(SuccessfulTask(self.0.clone())));
+        Ok(())
+    }
+}
+
+#[derive(Debug)]
+struct SuccessfulTask(Arc<Mutex<bool>>);
+
+#[async_trait::async_trait]
+impl Task for SuccessfulTask {
+    fn name(&self) -> &'static str {
+        "successful_task"
+    }
+    async fn run(self: Box<Self>, _stop_receiver: StopReceiver) -> anyhow::Result<()> {
+        let mut guard = self.0.lock().await;
+        *guard = true;
+        Ok(())
+    }
+}
+
+// Layer for task that updates resource and returns only after stop signal received
+#[derive(Debug)]
+struct RemainingTaskLayer(Arc<Mutex<bool>>);
+
+#[async_trait::async_trait]
+impl WiringLayer for RemainingTaskLayer {
+    fn layer_name(&self) -> &'static str {
+        "remaining_task_layer"
+    }
+
+    async fn wire(self: Box<Self>, mut node: ServiceContext<'_>) -> Result<(), WiringError> {
+        node.add_task(Box::new(RemainingTask(self.0.clone())));
+        Ok(())
+    }
+}
+
+#[derive(Debug)]
+struct RemainingTask(Arc<Mutex<bool>>);
+
+#[async_trait::async_trait]
+impl Task for RemainingTask {
+    fn name(&self) -> &'static str {
+        "remaining_task"
+    }
+    async fn run(self: Box<Self>, stop_receiver: StopReceiver) -> anyhow::Result<()> {
+        loop {
+            if *stop_receiver.0.borrow() {
+                let mut guard = self.0.lock().await;
+                *guard = false;
+                break;
+            }
+        }
```

Comment:
I guess you can get rid of the loop by using `stop_receiver.0.changed().await` before setting `guard` to `true`.